File: src/services/eventTranslationLayer.ts
import type { RealtimeEventData } from './eventStream';
import type { Workflow, TaskItem } from './workflowsApi';

export type UserTaskStatus =
  | 'ANALYZING'
  | 'EVALUATING'
  | 'SYNTHESIZING'
  | 'WAITING_FOR_APPROVAL'
  | 'EXECUTING'
  | 'COMPLETED'
  | 'FAILED'
  | 'DEGRADED';

export interface UserFacingTask {
  id: string;
  title: string;
  status: UserTaskStatus;
  progress: number;
  subSteps: string[];
  currentStep?: string;
  timestamp: string;
  summary?: string;
  details?: any;
}

const AGENT_TO_BUSINESS_TITLE: Record<string, string> = {
  BusinessAnalysisAgent: 'Business Performance Analysis',
  BusinessAnalysis: 'Business Performance Analysis',
  SEOAuditAgent: 'Website Performance Analysis',
  SEOAudit: 'Website Performance Analysis',
  CompetitorResearchAgent: 'Market Intelligence',
  CompetitorResearch: 'Market Intelligence',
  MarketingStrategyAgent: 'Strategic Planning',
  MarketingStrategy: 'Strategic Planning',
  CampaignPlannerAgent: 'Campaign Preparation',
  CampaignPlanner: 'Campaign Preparation',
  SwarmConsensus: 'Strategic Recommendation Validation',
  CEOAgent: 'Executive Alignment & Decisioning',
  PredictionAgent: 'Predictive Intelligence & Forecasting',
  PolicyEvolutionAgent: 'Policy & Governance Validation',
  ExecutionAgent: 'Autonomous Mission Execution'
};

export function mapAgentToBusinessTitle(agentName?: string): string {
  if (!agentName) return 'Business Intelligence Task';
  if (AGENT_TO_BUSINESS_TITLE[agentName]) {
    return AGENT_TO_BUSINESS_TITLE[agentName];
  }
  return agentName
    .replace(/Agent$/, '')
    .replace(/([A-Z])/g, ' $1')
    .trim();
}

export function translateBackendTaskToUserTask(task: TaskItem): UserFacingTask {
  const title = mapAgentToBusinessTitle(task.agent_name || task.title);

  let status: UserTaskStatus = 'ANALYZING';
  let progress = 0;
  const subSteps: string[] = ['Collecting verified business data', 'Evaluating current trends', 'Preparing strategic insights'];

  switch (task.status) {
    case 'RUNNING':
      status = 'ANALYZING';
      progress = 50;
      break;
    case 'COMPLETED':
      status = 'COMPLETED';
      progress = 100;
      break;
    case 'FAILED':
      status = 'FAILED';
      progress = 100;
      break;
    case 'QUEUED':
    case 'WAITING':
    default:
      status = 'EVALUATING';
      progress = 10;
      break;
  }

  return {
    id: task.id,
    title,
    status,
    progress,
    subSteps,
    currentStep: subSteps[0],
    timestamp: new Date().toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
    summary: `Task completed using verified business telemetry.`
  };
}

export function translateWorkflowToTasks(workflow: Workflow, tasks: TaskItem[] = []): {
  activeTask: UserFacingTask | null;
  completedTasks: UserFacingTask[];
  upcomingTasks: UserFacingTask[];
} {
  if (!workflow) {
    return { activeTask: null, completedTasks: [], upcomingTasks: [] };
  }

  const translated = tasks.map(t => translateBackendTaskToUserTask(t));

  const active = translated.find(t => t.status !== 'COMPLETED' && t.status !== 'FAILED') || null;
  const completed = translated.filter(t => t.status === 'COMPLETED');
  const upcoming = translated.filter(t => t !== active && t.status !== 'COMPLETED' && t.status !== 'FAILED');

  return {
    activeTask: active,
    completedTasks: completed,
    upcomingTasks: upcoming
  };
}

export function translateSSEEventToTaskUpdate(event: RealtimeEventData): Partial<UserFacingTask> {
  const title = mapAgentToBusinessTitle(event.agent_name);
  let status: UserTaskStatus = 'ANALYZING';

  if (event.event_type.includes('completed')) {
    status = 'COMPLETED';
  } else if (event.event_type.includes('failed')) {
    status = 'FAILED';
  } else if (event.event_type.includes('approval') || event.event_type.includes('pending')) {
    status = 'WAITING_FOR_APPROVAL';
  } else if (event.event_type.includes('execution')) {
    status = 'EXECUTING';
  }

  return {
    id: event.task_id || event.event_id,
    title,
    status,
    progress: event.progress || (status === 'COMPLETED' ? 100 : 50),
    summary: event.message || `Event processed by STRtOS Engine.`
  };
}

File: backend/app/services/event_translation_layer.py
from typing import Dict, Any, List, Optional

def extract_task_result_data(output: Any, title: str) -> Dict[str, Any]:
    if not output or not isinstance(output, dict):
        return {
            "title": title,
            "summary": "INSUFFICIENT DATA",
            "keyFinding": "INSUFFICIENT DATA",
            "importantChange": "INSUFFICIENT DATA",
            "businessImpact": "INSUFFICIENT DATA",
            "recommendation": "No recommendation available",
            "confidence": None,
            "findings": []
        }

    findings = output.get("findings") if isinstance(output.get("findings"), list) else []

    return {
        "title": title,
        "summary": output.get("summary") or (findings[0] if len(findings) > 0 else "INSUFFICIENT DATA"),
        "keyFinding": output.get("key_finding") or (findings[0] if len(findings) > 0 else "INSUFFICIENT DATA"),
        "importantChange": output.get("important_change") or (findings[1] if len(findings) > 1 else "INSUFFICIENT DATA"),
        "businessImpact": output.get("business_impact") or output.get("summary") or "INSUFFICIENT DATA",
        "recommendation": output.get("recommendation") or (output.get("recommendations")[0] if isinstance(output.get("recommendations"), list) and len(output.get("recommendations")) > 0 else "No recommendation available"),
        "confidence": output.get("confidence") if isinstance(output.get("confidence"), (int, float)) else None,
        "findings": findings
    }

def translate_backend_task_to_user_task(task: Any) -> Dict[str, Any]:
    status_raw = str(getattr(task, "status", "QUEUED")).upper()
    status = "QUEUED"
    
    if status_raw == "RUNNING":
        status = "RUNNING"
    elif status_raw == "COMPLETED":
        status = "COMPLETED"
    elif status_raw == "FAILED":
        status = "FAILED"
    elif status_raw in ["BLOCKED", "SKIPPED"]:
        status = "BLOCKED"

    status_msg = "StrtOS is working"
    if status == "COMPLETED":
        status_msg = "Completed successfully"
    elif status == "FAILED":
        status_msg = f"{getattr(task, 'title', 'Task')} could not be completed"
    elif status == "BLOCKED":
        status_msg = "Waiting for required analysis"

    output = getattr(task, "output", None)
    result = extract_task_result_data(output, getattr(task, "title", "Task")) if status == "COMPLETED" and output else None

    return {
        "id": getattr(task, "id", "t-1"),
        "workflowId": getattr(task, "workflow_id", "wf-1"),
        "title": getattr(task, "title", "Task"),
        "agentName": getattr(task, "agent_name", "Specialist Agent"),
        "status": status,
        "statusMessage": status_msg,
        "summary": result.get("summary") if result else ("INSUFFICIENT DATA" if status == "COMPLETED" else None),
        "errorReason": getattr(task, "error_message", None),
        "result": result
    }

def translate_workflow_to_tasks(workflow: Any, tasks: List[Any] = None) -> Dict[str, Any]:
    if not tasks:
        tasks = []
    
    translated = [translate_backend_task_to_user_task(t) for t in tasks]
    
    active = next((t for t in translated if t["status"] == "RUNNING"), None)
    completed = [t for t in translated if t["status"] == "COMPLETED"]
    upcoming = [t for t in translated if t["status"] in ["QUEUED", "BLOCKED"]]

    return {
        "activeTask": active,
        "completedTasks": completed,
        "upcomingTasks": upcoming
    }
